Firefly 667 Firefly related changes needed for irsakubernetes deployment

loitly opened the pull request:
https://jira.ipac.caltech.edu/browse/FIREFLY-667

- Firefly 'dev' and 'gwt' need to be modified so that it take header ''X-Forwarded-*' header' when available.  This is because when Firefly is behind a reverse proxy, the proxied host information is stored in these headers.
- Add health check endpoint to all Firefly-based applications.  The check will perform a simple operation, like allocating a small amount of memory to ensure the JVM is healthy.  It will be mapped to `/healthz`.
- Fixed error messages coming from the catalog dialog.

I do not have a test build.  Please review code to ensure it does not break existing functionality.  More testing to come when IRSA UI apps are served by irsakudev.

robyww commented on src/firefly/java/edu/caltech/ipac/firefly/server/servlets/HealthCheck.java:
```diff
@@ -0,0 +1,36 @@
+        Logger.debug(String.format("HealthCheck(Used/Max/Allocated): %s/%s/%s",
+                            FileUtil.getSizeAsString(rt.totalMemory() - rt.freeMemory()),
+                            FileUtil.getSizeAsString(rt.maxMemory()), FileUtil.getSizeAsString(b.length)));
+    }
```
You might want to catch an `OutOfMemoryError` here. and log that as well.

robyww replied:
I am guessing this is called outside of firefly app context. I would think it should return a string such a `Health Check Passed` or something.

You could also return a `Health Check Failed` for an `OutOfMemoryError` but I doubt it would return in the case. I have enchanted  edge case when we do get an`OutOfMemoryError` right before the server hangs.

loitly replied:
This is not meant for interactive use.  It's an HTTP API where a 200 status is pass and everything else is fail.
In the case of OOME or any exception, it will be caught upstream and a 500 will be returned.
This is usually used with a `timeout` threshold.  When it failed to return within a specified time, it is also considered 'unhealthy' or fail.

The `/healthz` check is within an app context.  So, for firefy, it would be `/firefly/healthz`.

robyww replied:
If you are logging it is still worth catching the `OOME`

loitly replied:
I can if you like, for logging purposes.  But in most cases, you won't even get an exception.  It will just hang there.

Normally, I don't want to log this.  It's unnecessary noise.  But, `debug` is our lowest log level.  This segue into our need to move most of our logs to 'info' level and set that as the default log level.

ejoliet replied:
the OOME is catch anyway in catalina.out, this is redundant to add it here in my opinion.

loitly replied:
I also think it's redundant, since the code will look like this.  However, I am fine with adding it. 
```
try { 
... 
} catch(oome) {
  log(oome);
  throw oome;
}
```

robyww commented on config/web.xml:
```diff
@@ -70,6 +70,15 @@
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>healthz</servlet-name>
+        <url-pattern>/healthz</url-pattern>
```
why `healthz` and not just `health`?

loitly replied:
> It historically comes from Google’s internal practices. They're called "z-pages".

I am following Kubernetes's convention which is from Google.  It is also popular in the kubernetes/docker community.
The 'z' is also to avoid any conflict with services endpoint.

robyww commented on src/firefly/java/edu/caltech/ipac/firefly/server/servlets/ServerStatus.java:
```diff
@@ -43,8 +48,7 @@
 
-        boolean showHeaders = Boolean.valueOf(req.getParameter("headers"));
-        String jwt = req.getParameter("jwt");
+        boolean showHeaders = Boolean.parseBoolean(req.getParameter("headers"));
```
This appears to be a special view of the server status page. why not just add it to everything else?

loitly replied:
Yes.  I added this a while back.   Displaying headers may take up lots of space that's why I did not add it by default.  Also, I am thinking we can expand on this and have additional parameters as needed in the future, like actions.

robyww reviewed (approved):
Looks fine. I have some comments below.

ejoliet reviewed (approved):
Had a question more about been curious than an issue so feel free to merge. The idea of updating the code to get the header specific values is ok in the code so all good.

loitly commented on src/firefly/js/tables/ui/TableRenderer.js:
```diff
@@ -406,19 +406,19 @@ export const LinkCell = React.memo(({cellInfo, style, ...rest}) => {
         } else {
             return (
-                <Cell {...CellProps}>
+                <Cell {...{rowIndex, height, width, columnKey}}>
```
As you mentioned, CellProps contains more than just these props.  However, `Cell` component is expecting only these and have PropTypes validation on these only.  That's it was giving validation errors.

loitly commented:
@robyww I added OOME logging to HealthCheck and fixed Description not making into the client `table.resources`.  Please take a look.  If no further request, I will merge this and all related PRs tomorrow.  Thanks for reviewing.

robyww commented:
Looks good. Fine with me to merge.
